File: services/alert-service/metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	metricAnomaliesConsumed = promauto.NewCounter(prometheus.CounterOpts{
		Name: "anomalies_consumed_total",
		Help: "Anomaly events consumed from anomalies.detected.",
	})

	metricAlertsGenerated = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "alerts_generated_total",
		Help: "Alerts created, by severity.",
	}, []string{"severity"})

	metricAlertsSuppressed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "alerts_suppressed_total",
		Help: "Alert-worthy events suppressed by dedup or cooldown, by reason.",
	}, []string{"reason"})

	metricAlertsEscalated = promauto.NewCounter(prometheus.CounterOpts{
		Name: "alerts_escalated_total",
		Help: "Alerts escalated to a higher severity after remaining unacknowledged.",
	})

	metricIncidentsOpen = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "incidents_open_total",
		Help: "Currently open alerts (proxy for open incidents until Phase 8 adds incident records).",
	})

	metricNotificationSent = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "notifications_sent_total",
		Help: "Notifications successfully sent, by provider.",
	}, []string{"provider"})

	metricNotificationFailed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "notifications_failed_total",
		Help: "Notification attempts that failed, by provider.",
	}, []string{"provider"})

	metricMessagesFailed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "messages_failed_total",
		Help: "Messages that failed processing, by reason.",
	}, []string{"reason"})

	metricDLQMessages = promauto.NewCounter(prometheus.CounterOpts{
		Name: "dlq_messages_total",
		Help: "Messages routed to the dead-letter topic.",
	})
)

File: services/alert-service/metrics.py
"""Prometheus metrics, the Python port of metrics.go. Same metric names and
labels, so existing Grafana dashboards/alerts keep working unchanged."""

from prometheus_client import Counter, Gauge

anomalies_consumed_total = Counter(
    "anomalies_consumed_total",
    "Anomaly events consumed from anomalies.detected.",
)

alerts_generated_total = Counter(
    "alerts_generated_total",
    "Alerts created, by severity.",
    ["severity"],
)

alerts_suppressed_total = Counter(
    "alerts_suppressed_total",
    "Alert-worthy events suppressed by dedup or cooldown, by reason.",
    ["reason"],
)

alerts_escalated_total = Counter(
    "alerts_escalated_total",
    "Alerts escalated to a higher severity after remaining unacknowledged.",
)

incidents_open_total = Gauge(
    "incidents_open_total",
    "Currently open (non-terminal) incidents.",
)

incidents_open_by_severity = Gauge(
    "incidents_open_by_severity",
    "Currently open (non-terminal) incidents, by severity -- what the Grafana 'critical incidents' panel actually needs.",
    ["severity"],
)

incidents_created_total = Counter(
    "incidents_created_total",
    "New incidents opened from an alert.",
)

alerts_attached_to_incident_total = Counter(
    "alerts_attached_to_incident_total",
    "Alerts attached to an already-open incident instead of creating a new one.",
)

notifications_sent_total = Counter(
    "notifications_sent_total",
    "Notifications successfully sent, by provider.",
    ["provider"],
)

notifications_failed_total = Counter(
    "notifications_failed_total",
    "Notification attempts that failed, by provider.",
    ["provider"],
)

messages_failed_total = Counter(
    "messages_failed_total",
    "Messages that failed processing, by reason.",
    ["reason"],
)

dlq_messages_total = Counter(
    "dlq_messages_total",
    "Messages routed to the dead-letter topic.",
)
